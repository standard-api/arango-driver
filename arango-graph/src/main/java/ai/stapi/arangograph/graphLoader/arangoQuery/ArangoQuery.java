package ai.stapi.arangograph.graphLoader.arangoQuery;

import ai.stapi.arangograph.graphLoader.arangoQuery.ast.AqlNode;
import ai.stapi.graphoperations.graphLoader.search.ResolvedQueryPart;
import com.fasterxml.jackson.annotation.JsonIgnore;
import java.util.Map;

public class ArangoQuery implements ResolvedQueryPart {

  private final AqlNode aqlNode;
  private final Map<String, Object> bindParameters;

  public ArangoQuery(AqlNode aqlNode, Map<String, Object> bindParameters) {
    this.aqlNode = aqlNode;
    this.bindParameters = bindParameters;
  }

  @JsonIgnore
  public AqlNode getAqlNode() {
    return aqlNode;
  }

  public String getQueryString() {
    return this.aqlNode.toQueryString();
  }

  public Map<String, Object> getBindParameters() {
    return bindParameters;
  }
}
