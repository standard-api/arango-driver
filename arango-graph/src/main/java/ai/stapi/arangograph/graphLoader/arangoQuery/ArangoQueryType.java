package ai.stapi.arangograph.graphLoader.arangoQuery;

public enum ArangoQueryType {
  NODE,
  OUTGOING_EDGE,
  INGOING_EDGE,
  GRAPH_TRAVERSAL
}
