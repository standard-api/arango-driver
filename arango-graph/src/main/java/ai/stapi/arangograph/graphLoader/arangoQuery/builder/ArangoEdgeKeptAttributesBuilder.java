package ai.stapi.arangograph.graphLoader.arangoQuery.builder;

import ai.stapi.arangograph.graphLoader.arangoQuery.ast.AqlVariable;
import ai.stapi.schema.structureSchemaProvider.StructureSchemaFinder;

public class ArangoEdgeKeptAttributesBuilder extends ArangoKeptAttributesBuilder {

  protected static final String[] DEFAULT_META_ATTRIBUTES = {
      "_id",
      "_rev",
      "_key",
      "_from",
      "_to",
      "_metaData"
  };

  public ArangoEdgeKeptAttributesBuilder(
      StructureSchemaFinder structureSchemaProvider,
      AqlVariable documentName,
      String subQueryPostfix,
      String graphElementType
  ) {
    super(structureSchemaProvider, documentName, subQueryPostfix, graphElementType);
  }

  @Override
  protected String[] provideDefaultMetaAttributes() {
    return ArangoEdgeKeptAttributesBuilder.DEFAULT_META_ATTRIBUTES;
  }

}
