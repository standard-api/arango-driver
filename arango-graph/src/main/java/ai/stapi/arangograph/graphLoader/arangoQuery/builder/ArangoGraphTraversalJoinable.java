package ai.stapi.arangograph.graphLoader.arangoQuery.builder;

import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.AbstractEdgeDescription;

public interface ArangoGraphTraversalJoinable {

  ArangoGraphTraversalSubQueryBuilder joinGraphTraversal(AbstractEdgeDescription edgeDescription);

  ArangoGraphTraversalSubQueryBuilder mapGraphTraversal(
      String fieldName,
      AbstractEdgeDescription edgeDescription
  );

  ArangoGraphTraversalSubQueryBuilder mapGraphTraversalAsConnections(
      String fieldName,
      AbstractEdgeDescription edgeDescription
  );

}
