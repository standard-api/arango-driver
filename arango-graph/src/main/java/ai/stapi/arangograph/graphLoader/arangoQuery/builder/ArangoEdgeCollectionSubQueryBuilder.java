package ai.stapi.arangograph.graphLoader.arangoQuery.builder;

import ai.stapi.arangograph.graphLoader.arangoQuery.ArangoQuery;
import ai.stapi.arangograph.graphLoader.arangoQuery.ast.AqlObject;
import ai.stapi.arangograph.graphLoader.arangoQuery.ast.highLevelOperations.AqlFor;
import ai.stapi.arangograph.graphLoader.arangoQuery.searchOptionResolvers.ArangoGenericSearchOptionResolver;
import ai.stapi.graphoperations.graphLoader.GraphLoaderReturnType;
import ai.stapi.schema.structureSchemaProvider.StructureSchemaFinder;
import org.jetbrains.annotations.NotNull;

public class ArangoEdgeCollectionSubQueryBuilder extends ArangoEdgeSubQueryBuilder {

  private ArangoEdgeCollectionSubQueryBuilder(
      ArangoGenericSearchOptionResolver searchOptionResolver,
      StructureSchemaFinder structureSchemaFinder,
      String graphElementType,
      String subQueryPostfix,
      boolean isOutgoing
  ) {
    super(searchOptionResolver, structureSchemaFinder, graphElementType, subQueryPostfix,
        isOutgoing);
  }

  public static ArangoEdgeCollectionSubQueryBuilder ingoing(
      ArangoGenericSearchOptionResolver searchOptionResolver,
      StructureSchemaFinder structureSchemaFinder,
      String graphElementType,
      String subQueryPostfix
  ) {
    return new ArangoEdgeCollectionSubQueryBuilder(
        searchOptionResolver,
        structureSchemaFinder,
        graphElementType,
        subQueryPostfix,
        false
    );
  }

  public static ArangoEdgeCollectionSubQueryBuilder outgoing(
      ArangoGenericSearchOptionResolver searchOptionResolver,
      StructureSchemaFinder structureSchemaFinder,
      String graphElementType,
      String subQueryPostfix
  ) {
    return new ArangoEdgeCollectionSubQueryBuilder(
        searchOptionResolver,
        structureSchemaFinder,
        graphElementType,
        subQueryPostfix,
        true
    );
  }

  @Override
  @NotNull
  protected ArangoQuery createAqlBody(
      ArangoQuery keptAttributes,
      AqlObject graph,
      GraphLoaderReturnType... returnTypes
  ) {
    var collectionVariable =
        this.createCollectionVariable(this.subQueryPostfix).markAsBindParameter();
    var aqlFor = new AqlFor(
        this.createEdgeDocumentName(this.subQueryPostfix),
        collectionVariable.markAsBindParameter()
    );
    var aqlBody = this.createBaseAqlBody(keptAttributes, graph, aqlFor, returnTypes);
    aqlBody.getBindParameters().put(collectionVariable.getVariableName(), this.graphElementType);
    return aqlBody;
  }

}
