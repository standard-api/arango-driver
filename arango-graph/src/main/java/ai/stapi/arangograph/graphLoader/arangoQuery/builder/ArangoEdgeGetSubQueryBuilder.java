package ai.stapi.arangograph.graphLoader.arangoQuery.builder;

import ai.stapi.arangograph.graphLoader.arangoQuery.ArangoQuery;
import ai.stapi.arangograph.graphLoader.arangoQuery.ast.AqlObject;
import ai.stapi.arangograph.graphLoader.arangoQuery.ast.AqlString;
import ai.stapi.arangograph.graphLoader.arangoQuery.ast.functions.AqlConcatSeparator;
import ai.stapi.arangograph.graphLoader.arangoQuery.ast.functions.AqlDocument;
import ai.stapi.arangograph.graphLoader.arangoQuery.ast.highLevelOperations.AqlLet;
import ai.stapi.arangograph.graphLoader.arangoQuery.searchOptionResolvers.ArangoGenericSearchOptionResolver;
import ai.stapi.graphoperations.graphLoader.GraphLoaderReturnType;
import ai.stapi.identity.UniqueIdentifier;
import ai.stapi.schema.structureSchemaProvider.StructureSchemaFinder;
import org.jetbrains.annotations.NotNull;

public class ArangoEdgeGetSubQueryBuilder extends ArangoEdgeSubQueryBuilder {

  private final UniqueIdentifier graphElementId;

  public ArangoEdgeGetSubQueryBuilder(
      ArangoGenericSearchOptionResolver searchOptionResolver,
      StructureSchemaFinder structureSchemaFinder,
      String graphElementType,
      String subQueryPostfix,
      boolean isOutgoing,
      UniqueIdentifier graphElementId
  ) {
    super(searchOptionResolver, structureSchemaFinder, graphElementType, subQueryPostfix,
        isOutgoing);
    this.graphElementId = graphElementId;
  }

  public static ArangoEdgeGetSubQueryBuilder ingoing(
      ArangoGenericSearchOptionResolver searchOptionResolver,
      StructureSchemaFinder structureSchemaFinder,
      UniqueIdentifier graphElementUuid,
      String graphElementType,
      String subQueryPostfix
  ) {
    return new ArangoEdgeGetSubQueryBuilder(
        searchOptionResolver,
        structureSchemaFinder,
        graphElementType,
        subQueryPostfix,
        false,
        graphElementUuid
    );
  }

  public static ArangoEdgeGetSubQueryBuilder outgoing(
      ArangoGenericSearchOptionResolver searchOptionResolver,
      StructureSchemaFinder structureSchemaFinder,
      UniqueIdentifier graphElementUuid,
      String graphElementType,
      String subQueryPostfix
  ) {
    return new ArangoEdgeGetSubQueryBuilder(
        searchOptionResolver,
        structureSchemaFinder,
        graphElementType,
        subQueryPostfix,
        true,
        graphElementUuid
    );
  }

  @NotNull
  @Override
  protected ArangoQuery createAqlBody(
      ArangoQuery keptAttributes,
      AqlObject graph,
      GraphLoaderReturnType... returnTypes
  ) {
    var collectionVariable = this.createCollectionVariable(this.subQueryPostfix);
    var idVariable = this.createIdVariable(this.subQueryPostfix);
    var aqlFor = new AqlLet(
        this.createEdgeDocumentName(this.subQueryPostfix),
        new AqlDocument(
            new AqlConcatSeparator(
                new AqlString("/"),
                collectionVariable.markAsBindParameter(),
                idVariable.markAsBindParameter()
            )
        )
    );
    var baseAqlBody = this.createBaseAqlBody(keptAttributes, graph, aqlFor, returnTypes);
    baseAqlBody.getBindParameters()
        .put(collectionVariable.getVariableName(), this.graphElementType);
    baseAqlBody.getBindParameters().put(idVariable.getVariableName(), this.graphElementId.getId());
    return baseAqlBody;
  }
}
