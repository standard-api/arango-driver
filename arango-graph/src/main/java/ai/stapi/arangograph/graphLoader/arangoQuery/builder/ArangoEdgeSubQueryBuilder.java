package ai.stapi.arangograph.graphLoader.arangoQuery.builder;

import ai.stapi.arangograph.graphLoader.arangoQuery.ArangoQuery;
import ai.stapi.arangograph.graphLoader.arangoQuery.ArangoQueryType;
import ai.stapi.arangograph.graphLoader.arangoQuery.ast.AqlNode;
import ai.stapi.arangograph.graphLoader.arangoQuery.ast.AqlObject;
import ai.stapi.arangograph.graphLoader.arangoQuery.ast.AqlRootNode;
import ai.stapi.arangograph.graphLoader.arangoQuery.ast.AqlString;
import ai.stapi.arangograph.graphLoader.arangoQuery.ast.AqlVariable;
import ai.stapi.arangograph.graphLoader.arangoQuery.ast.functions.AqlDocument;
import ai.stapi.arangograph.graphLoader.arangoQuery.ast.functions.AqlPush;
import ai.stapi.arangograph.graphLoader.arangoQuery.ast.highLevelOperations.AqlLet;
import ai.stapi.arangograph.graphLoader.arangoQuery.searchOptionResolvers.ArangoGenericSearchOptionResolver;
import ai.stapi.graphoperations.graphLoader.GraphLoaderReturnType;
import ai.stapi.schema.structureSchemaProvider.StructureSchemaFinder;
import java.util.Map;
import org.jetbrains.annotations.NotNull;

public abstract class ArangoEdgeSubQueryBuilder implements ArangoMainQueryBuilder {

  protected final boolean isOutgoing;
  protected final ArangoGenericSearchOptionResolver searchOptionResolver;
  protected final StructureSchemaFinder structureSchemaFinder;
  protected final ArangoQueryByNodeTypeBuilder arangoQueryByNodeTypeBuilder;
  protected final ArangoEdgeKeptAttributesBuilder arangoEdgeKeptAttributesBuilder;
  protected final ArangoMappedObjectBuilder arangoMappedObjectBuilder;
  protected final ArangoSearchOptionsBuilder arangoSearchOptionsBuilder;
  protected final String graphElementType;
  protected final String subQueryPostfix;
  private boolean isNodeFieldMapped = false;

  public ArangoEdgeSubQueryBuilder(
      ArangoGenericSearchOptionResolver searchOptionResolver,
      StructureSchemaFinder structureSchemaFinder,
      String graphElementType,
      String subQueryPostfix,
      boolean isOutgoing
  ) {
    this.searchOptionResolver = searchOptionResolver;
    this.structureSchemaFinder = structureSchemaFinder;
    this.graphElementType = graphElementType;
    this.subQueryPostfix = subQueryPostfix;
    this.isOutgoing = isOutgoing;
    this.arangoQueryByNodeTypeBuilder = new ArangoQueryByNodeTypeBuilder(
        this.searchOptionResolver,
        this.structureSchemaFinder,
        this.createChildSubQueryPostfix(this.subQueryPostfix, 0),
        this.buildJoinedNodeGetter()
    );
    var edgeDocumentName = this.createEdgeDocumentName(this.subQueryPostfix);
    this.arangoEdgeKeptAttributesBuilder = new ArangoEdgeKeptAttributesBuilder(
        this.structureSchemaFinder,
        edgeDocumentName,
        this.subQueryPostfix,
        this.graphElementType
    );
    this.arangoMappedObjectBuilder = new ArangoMappedObjectBuilder(
        this.arangoEdgeKeptAttributesBuilder,
        this.structureSchemaFinder,
        edgeDocumentName,
        this.subQueryPostfix,
        this.graphElementType
    );
    this.arangoSearchOptionsBuilder = new ArangoSearchOptionsBuilder(
        this.searchOptionResolver,
        edgeDocumentName,
        this.isOutgoing ? ArangoQueryType.OUTGOING_EDGE : ArangoQueryType.INGOING_EDGE,
        this.graphElementType,
        this.subQueryPostfix
    );
  }

  public ArangoEdgeKeptAttributesBuilder setKeptAttributes() {
    return this.arangoEdgeKeptAttributesBuilder;
  }

  public ArangoMappedObjectBuilder setMappedScalars() {
    return this.arangoMappedObjectBuilder;
  }

  public ArangoSearchOptionsBuilder setSearchOptions() {
    return this.arangoSearchOptionsBuilder;
  }

  public ArangoNodeGetSubQueryBuilder joinNodeGetSubQuery(String nodeType) {
    return this.arangoQueryByNodeTypeBuilder.addGetNodeOption(nodeType);
  }

  public ArangoNodeGetSubQueryBuilder mapNodeGetSubQuery(String nodeType) {
    if (!this.isNodeFieldMapped) {
      this.arangoMappedObjectBuilder.mapCustomField(
          "node",
          this.createSubQueryVariable(this.createChildSubQueryPostfix(this.subQueryPostfix, 0))
              .getItem(0)
              .getField("data")
      );
      this.isNodeFieldMapped = true;
    }
    return this.arangoQueryByNodeTypeBuilder.addGetNodeOption(nodeType);
  }

  @Override
  public ArangoQuery build(GraphLoaderReturnType... returnTypes) {
    var subQueryVariable =
        this.createSubQueryVariable(this.createChildSubQueryPostfix(this.subQueryPostfix, 0));
    var keptAttributes = this.arangoEdgeKeptAttributesBuilder.build();
    var edges = this.createEdges(subQueryVariable, keptAttributes);
    var nodes = this.createNodes(subQueryVariable);
    var graph = this.createGraphObject(nodes, edges);

    return this.createAqlBody(keptAttributes, graph, returnTypes);
  }

  @Override
  public ArangoQuery buildAsMain(GraphLoaderReturnType... returnTypes) {
    var subQueryVariable =
        this.createSubQueryVariable(this.createChildSubQueryPostfix(this.subQueryPostfix, 0));
    var keptAttributes = this.arangoEdgeKeptAttributesBuilder.build();
    var main = keptAttributes.getAqlNode();
    var nodes = this.createNodes(subQueryVariable);
    var edges = this.getChildSubQueryEdges(subQueryVariable);
    var graph = this.createMainGraphObject(main, nodes, edges);

    return this.createAqlBody(keptAttributes, graph, returnTypes);
  }

  @NotNull
  protected abstract ArangoQuery createAqlBody(
      ArangoQuery keptAttributes,
      AqlObject graph,
      GraphLoaderReturnType... returnTypes
  );

  @NotNull
  protected ArangoQuery createBaseAqlBody(
      ArangoQuery keptAttributes,
      AqlObject graph,
      AqlNode getDocumentLine,
      GraphLoaderReturnType... returnTypes
  ) {
    var mappedObject = this.arangoMappedObjectBuilder.build();
    var dataObject = new AqlObject(Map.of(
        new AqlString("edges"), mappedObject.getAqlNode()
    ));
    var childQuery = this.arangoQueryByNodeTypeBuilder.build(returnTypes);
    var searchOptions = this.arangoSearchOptionsBuilder.build();

    var letJoinedNode = new AqlLet(
        this.createSubQueryVariable(this.createChildSubQueryPostfix(this.subQueryPostfix, 0)),
        childQuery.getAqlNode()
    );

    var forQuery = new AqlRootNode(
        getDocumentLine,
        searchOptions.getAqlNode(),
        letJoinedNode,
        this.createOtherNodeLine(),
        this.createReturnStatement(
            graph,
            dataObject,
            returnTypes
        )
    );

    var bindParameters = this.mergeBindParameters(
        keptAttributes.getBindParameters(),
        mappedObject.getBindParameters(),
        childQuery.getBindParameters(),
        searchOptions.getBindParameters()
    );

    return new ArangoQuery(
        forQuery,
        bindParameters
    );
  }

  private AqlNode createEdges(AqlVariable subQueryVariable, ArangoQuery keptAttributes) {
    return new AqlPush(
        this.getChildSubQueryEdges(subQueryVariable),
        keptAttributes.getAqlNode()
    );
  }

  private AqlNode createNodes(AqlVariable subQueryVariable) {
    return new AqlPush(
        this.getChildSubQueryNodes(subQueryVariable),
        new ArangoNodeKeptAttributesBuilder(
            this.structureSchemaFinder,
            this.createOtherNodeVariableName(),
            "",
            ""
        ).build().getAqlNode()
    );
  }

  protected AqlVariable buildJoinedNodeGetter() {
    return this.createEdgeDocumentName(this.subQueryPostfix)
        .getField(this.isOutgoing ? "_to" : "_from");
  }

  private AqlLet createOtherNodeLine() {
    return new AqlLet(
        this.createOtherNodeVariableName(),
        new AqlDocument(
            this.createEdgeDocumentName(this.subQueryPostfix)
                .getField(this.isOutgoing ? "_from" : "_to")
        )
    );
  }

  @NotNull
  private AqlVariable createOtherNodeVariableName() {
    var childSubQueryPrefix = this.createChildSubQueryPostfix(this.subQueryPostfix, 0);
    return new AqlVariable("otherNode_" + childSubQueryPrefix);
  }
}
