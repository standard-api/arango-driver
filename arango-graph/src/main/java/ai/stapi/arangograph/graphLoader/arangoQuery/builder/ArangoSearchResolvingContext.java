package ai.stapi.arangograph.graphLoader.arangoQuery.builder;

import ai.stapi.arangograph.graphLoader.arangoQuery.ArangoQueryType;
import ai.stapi.arangograph.graphLoader.arangoQuery.ast.AqlVariable;
import ai.stapi.graphoperations.graphLoader.search.SearchResolvingContext;

public class ArangoSearchResolvingContext implements SearchResolvingContext {

  private final AqlVariable documentName;
  private final ArangoQueryType originQueryType;

  private final String graphElementType;
  private final String placeholderPostfix;
  private final String subQueryPostfix;

  public ArangoSearchResolvingContext(
      AqlVariable documentName,
      ArangoQueryType originQueryType,
      String graphElementType, String placeholderPostfix,
      String subQueryPostfix
  ) {
    this.documentName = documentName;
    this.originQueryType = originQueryType;
    this.graphElementType = graphElementType;
    this.placeholderPostfix = placeholderPostfix;
    this.subQueryPostfix = subQueryPostfix;
  }

  public ArangoSearchResolvingContext(
      String documentName,
      ArangoQueryType originQueryType,
      String graphElementType,
      String placeholderPostfix,
      String subQueryPostfix
  ) {
    this(new AqlVariable(documentName), originQueryType, graphElementType, placeholderPostfix,
        subQueryPostfix);
  }

  public ArangoSearchResolvingContext(AqlVariable documentName, ArangoQueryType originQueryType,
      String graphElementType) {
    this(documentName, originQueryType, graphElementType, "", "");
  }

  public ArangoSearchResolvingContext(String documentName, ArangoQueryType originQueryType,
      String graphElementType) {
    this(documentName, originQueryType, graphElementType, "", "");
  }

  public String getPlaceholderPostfix() {
    return placeholderPostfix;
  }

  public AqlVariable getDocumentName() {
    return documentName;
  }

  public String getSubQueryPostfix() {
    return subQueryPostfix;
  }

  public ArangoQueryType getOriginQueryType() {
    return originQueryType;
  }

  public String getGraphElementType() {
    return graphElementType;
  }
}
