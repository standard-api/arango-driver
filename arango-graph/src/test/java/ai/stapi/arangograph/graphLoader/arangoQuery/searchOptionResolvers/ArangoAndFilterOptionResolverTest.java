package ai.stapi.arangograph.graphLoader.arangoQuery.searchOptionResolvers;

import ai.stapi.arangograph.graphLoader.arangoQuery.ArangoQueryType;
import ai.stapi.arangograph.graphLoader.arangoQuery.builder.ArangoSearchResolvingContext;
import ai.stapi.arangograph.graphLoader.arangoQuery.searchOptionResolvers.exampleImplementations.ArangoFilterOptionResolverTestStructureLoader;
import ai.stapi.graphoperations.graphLoader.search.filterOption.AndFilterOption;
import ai.stapi.graphoperations.graphLoader.search.filterOption.EqualsFilterOption;
import ai.stapi.graphoperations.graphLoader.search.filterOption.GreaterThanFilterOption;
import ai.stapi.objectRenderer.infrastructure.objectToJsonStringRenderer.ObjectToJSonStringOptions;
import ai.stapi.test.schemaintegration.SchemaIntegrationTestCase;
import ai.stapi.test.schemaintegration.StructureDefinitionScope;
import java.util.List;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;

@StructureDefinitionScope(ArangoFilterOptionResolverTestStructureLoader.SCOPE)
public class ArangoAndFilterOptionResolverTest extends SchemaIntegrationTestCase {

  @Autowired
  private ArangoAndFilterOptionResolver resolver;

  @Test
  public void itShouldSupportAndFilterOption() {
    var fitler1 = new EqualsFilterOption<>("example_attribute_1", "Example Equal Value");
    var fitler2 = new GreaterThanFilterOption<>("example_attribute_2", 5);
    var andFilter = new AndFilterOption(List.of(fitler1, fitler2));
    var resolver = this.resolver;
    Assertions.assertTrue(
        resolver.supports(andFilter),
        "AndFilterOptionResolver should support AndFilterOption."
    );
  }

  @Test
  public void itShouldResolveOptionOfOne() {
    var fitler1 = new EqualsFilterOption<>("example_attribute_1", "Example Equal Value");
    var andFilter = new AndFilterOption(List.of(fitler1));
    var resolved = this.resolver.resolve(
        andFilter,
        new ArangoSearchResolvingContext("exampleDocumentName", ArangoQueryType.NODE,
            "graphElementType")
    );
    this.thenObjectApproved(resolved, this.getOptions());
  }

  @Test
  public void itShouldResolveOptionOfTwo() {
    var fitler1 = new EqualsFilterOption<>("example_attribute_1", "Example Equal Value");
    var fitler2 = new GreaterThanFilterOption<>("example_attribute_2", 5);
    var andFilter = new AndFilterOption(List.of(fitler1, fitler2));
    var resolved = this.resolver.resolve(
        andFilter,
        new ArangoSearchResolvingContext("exampleDocumentName", ArangoQueryType.NODE,
            "graphElementType")
    );
    this.thenObjectApproved(resolved, this.getOptions());
  }

  private ObjectToJSonStringOptions getOptions() {
    return new ObjectToJSonStringOptions(
        ObjectToJSonStringOptions.RenderFeature.SORT_FIELDS,
        ObjectToJSonStringOptions.RenderFeature.RENDER_GETTERS
    );
  }
}
