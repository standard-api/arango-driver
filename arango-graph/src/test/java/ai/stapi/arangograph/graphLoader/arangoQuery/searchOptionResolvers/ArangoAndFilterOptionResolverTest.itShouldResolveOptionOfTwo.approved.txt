{
  "queryString" : "( ( exampleDocumentName.attributes.@filterAttributeNamePlaceholder_and_0[0].values[*].value ) == @filterAttributeValuePlaceholder_and_0 AND ( exampleDocumentName.attributes.@filterAttributeNamePlaceholder_and_1[0].values[*].value ) > @filterAttributeValuePlaceholder_and_1 )",
  "bindParameters" : {
    "filterAttributeNamePlaceholder_and_0" : "example_attribute_1",
    "filterAttributeNamePlaceholder_and_1" : "example_attribute_2",
    "filterAttributeValuePlaceholder_and_0" : "Example Equal Value",
    "filterAttributeValuePlaceholder_and_1" : 5
  }
}